Домашнее задание по теме "Хитрости тестирования"

emvmikhailevg opened the pull request.

emvmikhailevg committed:
Все три задачи

vpyzhyanov commented on src/test/java/bot/BotLogicTest.java:
```diff
@@ -0,0 +1,137 @@
+
+    /**
+     * Тестирование команды "/test", если пользователь ответил на вопросы верно. Состояние должно
+     * поменяться на TEST, а бот должен отправить свой первый вопрос
```
Тестирование внутренней логики уместно только тогда, когда вы используете Mockito и ограничиваетесь тестированием только одного класса. Здесь же это не так. Состояние - внутренняя логика, по хорошему, вы даже не должны знать о её существовании. В общем проверять надо только корректность ответов бота и корректную реакцию на разные команды.

vpyzhyanov commented on src/test/java/bot/BotLogicTest.java:
```diff
@@ -0,0 +1,137 @@
+        Thread.sleep(1000);
+
+        // Из презентации ясно, что нужно написать эту строчку кода, но на ней вылетает
+        // исключение. Можете объяснить, обязательно ли ее писать в таком случае?
```
Комментарии в коде - не место для вопросов и обсуждений. Для этого есть телеграмм или на худой конец Pull Request.

Вопрос построен неправильно. 

1. Сначала надо описать проблему: какое именно исключение?
2. Надо самостоятельно попытаться разобраться в проблеме: Почему происходит исключение? В чём причина?
3. Предложить варианты решения и обосновать их

Вы же пропустили первые 2 пункта и сразу предложили какое-то решение. На основании текущих данных нельзя сделать вывод и ответить на ваш вопрос. Пока очевидно только то, что вы хотите удалить проверку только потому, что столкнулись с проблемой, разобраться в которой даже не попытались.

Я могу предположить что тут не так, но не буду этого делать, подожду ваш анализ проблемы.

emvmikhailevg replied:
На этой строчке кода вылетает исключение IndexOutOfBoundsException. Длина списка с сообщениями от бота равняется 3, я же в этой строчке обращаюсь к 4 элементу коллекции, из-за чего исключение и выбрасывается. Таким образом, у меня возник вопрос: стоит ли вообще писать эту проверку? (в презентации она была, здесь, наверное, тоже должна быть)

vpyzhyanov replied:
Написали, что она должна быть и при этом вы решили её убрать( Анализ проблемы до сих пор не полный. Надо искать настоящую причину, а не промежуточную. Надо задать вопрос "А почему в списке сообщений только 3 сообщения, почему 4 не появилось?". Делаю вывод, что вы не понимаете что переписываете из презентации, а бездумно копируете.

![image](https://github.com/emvmikhailevg/LessonTestTask/assets/71599019/e2711416-9f76-4401-b92d-ffeafd25d5d1)

Обратите внимание какое ожидание вы вводите (3 секунды) и какую задержку вы выставляете (1 секунда).

**То, что вы не можете найти причину ошибки не может являются основанием удалять проверки**

vpyzhyanov commented on src/test/java/bot/BotLogicTest.java:
```diff
@@ -0,0 +1,137 @@
+        botLogic.processCommand(user, "6");
+        Assert.assertEquals("Правильный ответ!", fakeBot.getMessages().get(9));
+        Assert.assertEquals(State.INIT, user.getState());
+        Assert.assertEquals("Тест завершен", fakeBot.getMessages().get(10));
```
тест не полный. Не хватает проверок

vpyzhyanov replied:
Не исправлено. Обратите внимание, что замечание написано к последней строчке, и имеет текст **"тест не полный"**, что означает, что вы его не закончили.

Ваши проверки добавленные в середину логически правильные, но они проверяют опять же внутреннюю логику. Вместо них должен быть аналог, без использования обращения к внутренней логике

vpyzhyanov commented on src/test/java/container/ContainerTest.java:
```diff
@@ -0,0 +1,86 @@
+     * Тестирование получения вещи из контейнера по индексу
+     */
+    @Test
+    public void getMethodTest() {
```
разве в задании было необходимо тестировать методы get и size?

emvmikhailevg replied:
Согласен, проверки были излишни, исправил.

vpyzhyanov commented on src/main/java/example/container/Item.java:
```diff
@@ -6,7 +6,7 @@
     private final long id;
 
-    Item(long id) {
+    public Item(long id) {
```
кто разрешил менять основной код? Убрать это!

emvmikhailevg replied:
Дело в том, что, если не сделать конструктор public, то в тестах его нельзя будет вызвать.
![image](https://github.com/emvmikhailevg/LessonTestTask/assets/144726521/08e0899a-460c-41e9-bd65-14b812b1fd66)

vpyzhyanov replied:
Убрать - означает **убрать** и ничего больше! Правильно было задать вопрос и преодолеть проблему правильным способом, а не первым попавшимся.

Всё дело в пакете, в котором вы создали тест. Если тест находится в таком же пакете, что и основной класс, то "пакетный доступ" позволяет нам обратиться к конструктору без проблем.

vpyzhyanov reviewed (changes requested):
Мало тестов на бота.

emvmikhailevg committed:
1. Убрал проверку состояния
2. Добавил проверку на количество неверных ответов юзера в методе repeatCommandTest
3. Убрал тестирование методов get и size

vpyzhyanov commented on src/test/java/bot/BotLogicTest.java:
```diff
@@ -80,26 +74,17 @@ public void testCommandTestForWrongAnswers() {
-        Assert.assertEquals(State.SET_NOTIFY_DELAY, user.getState());
         Assert.assertEquals("Через сколько секунд напомнить?", fakeBot.getMessages().get(1));
 
         // следует проверить, что пользователь введет именно число
```
комментарий есть, а проверки нет

vpyzhyanov reviewed (changes requested):
не все замечания были исправлены. Плохо.
